### ISSUE TITLE
File/directory names with [ ] always transcoded

### ISSUE THREAD

**Issue by bitchkat** [smxi/acxi#5]
Another weird thing I noticed is that if a directory or file namehas square brackets in the name, flac files are always transcoded.

So if a track is stored in file "01 -Song Title [year].flac"  It will be transcoded every time axci is run.  If I rename the file or directory to not have square brackets, it behaves normally.

**Comment by smxi** [comment]
oh, that means there is some regex somewhere that's using that [...] as a pattern, I thought I'd cleaned up all those. In the past I'd neglected to be aware of s/\Q$VAR\E/ in perl which does not treat regex characcters in $VAR as regex, I went through and thought I'd caught all of those, but I guess I missed one. Let me see if I can find it.

**Comment by smxi** [comment]
Hmm, that's odd, I just tested this and made a file with [letters] in it, transcoded first time, then not after. So my guess there was wrong about what's going on. I have a test directory with all the possible mutations I come across in terms of syntax i don't use, so I expected this to just work as it does in your case, but it didn't, or rather, it did, and did what it was supposed to, yet didn't in your case, which makes it harder to figure out.

**Comment by smxi** [comment]
hmm, [1984] is the actual value, let me test that.

**Comment by smxi** [comment]
No, I can't make it happen, created a file name with [1984] in it, and it transcoded first time as expected, then not again. So I'm stumped there. What OS and file system is this?

**Comment by smxi** [comment]
As an aside, I had a media player that I assume ran some linux variant under the covers, or not, don't know, and it would choke on > in a file name, which is an illegal windows file/path character. That was odd since > is allowed in linux/bsd as far as I know, and it certainly was not running windows. So I'm going to guess the filesystem type or os matters here.

**Comment by bitchkat** [comment]
Fedora 30, ext4 for both source and dest, perl 5.28.2

**Comment by smxi** [comment]
Then that's decidedly odd, since I'm testing on ext4 as well. Normally these things would reproduce, that's how I've debugged all odd file / path name character issues, just making them, and it trips the error, but in this case, it did not. Could this be a kernel security feature?

**Comment by smxi** [comment]
13 - Untitled [Outtake from (I'm) Stranded].flac type syntax tripped the exact opposite error, the system simply does not see the file name somehow.

**Comment by smxi** [comment]
Nope, with that exact file name, it processes it the first time, then sees it as processed subsequently. 

I think there's some difference going on which will be tricky to figure out, it certainly seems like something is interpreting that as regex on your system, but why? perl 5.28.1 here, nothing like that would have changed internally in perl.

**Comment by smxi** [comment]
I'm wondering/suspecting this may be related to some SE Linux setting, which I believe Fedora ships with?

**Comment by bitchkat** [comment]
I'll test tonight.  I was maybe a week to 10 days out of date on acxi if there was something checked in that may have affected this behavior. I updated yesterday but had already removed all  '[]' from directory and file names.  I added a config to kid3 to change '[' -> '(' and ']' -> ')' when renaming files.

**Comment by smxi** [comment]
You might want to discover why that's happening, given that we're running essentially the same setup, the only difference I can think of is SE Linux interfering and deciding a valid file path that contains regex characters is regex. That's just a guess, but it would be good to know the actual why of it. [ ] are not banned file path characters in Linux, in fact, I checked, and only \ is not allowed in a file name.

**Comment by bitchkat** [comment]
Yeah, if '[]' was not allowed in file names then my issue wouldn't have made any sense.  ;-)  I definitely have (or rather had) directories and files with '[text]' in the file names.  The SE Linux theory doesn't make sense since acxi is able to a.  read the source file with '[text]' in the file name and b. create and write to the file on dest dir with the extension changed from .flac to .ogg.  

I assume that once you find a source file, you generate the path the destination and then stat(2) both files and check if the dest is older than the src?  Or maybe you use cached checksums.  

Plus I generally run with "setenforce 0"  because it blocks drupal running in my home directory from doing some things.

**Comment by smxi** [comment]
Yes, I stat the source and destination file, or rather, acxi does. I never use checksums in that way.

SE Linux was just a guess since it's the only difference that I can readily spot, and it's often been commented that configuring it well is so complicated that many users ended up just disabling it, which is why I leaned towards suspecting that. Never used it myself though, so no firsthand knowledge.

All the file/path name processing in acxi is using the s/\Q$FILEPATH\E/ filter method, which is why I can't reproduce it. that bug was part of the big wave of fixes that I discovered last month, so it's been fixed. But it's still somewhat likely that what is happening is that somewhere in your system, something is interpreting that regex like string as regex anyway. Way above my pay grade to figure that one out, since if it were an acxi error, feeding in your literal file name would have tripped it immediately, which would mean it would be fixed now.
